File: code/mergeSort.cpp
#include<bits/stdc++.h>
using namespace std ;

void merge(int arr[], int l, int m , int r){
	int n1 = m-l+1;
	int n2 = r-m ;
	int L[n1];
	int R[n2];
	for (int i = 0; i < n1; i++)
        L[i] = arr[l + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[m + 1+ j];
	int i =0, j= 0 , k=l ;
	while(i<n1 && j<n2){
		if(L[i]<R[j]){
			arr[k]= L[i];
			i++;
		}
		else{
			arr[k]= R[j];
			j++;
		}
		k++;
	}
	while(i<n1){
		arr[k] = L[i];
		i++;k++;
	}
	while(j<n2){
		arr[k]= R[j];
		j++;k++;
	}
	
}
void mergeSort(int arr[], int l, int r){
	if(l<r){
		int m = l+ (r-l)/2 ;
		mergeSort(arr,l, m);
		mergeSort(arr,m+1,r);
		
		merge(arr, l,m,r);
	}
}
void printArray(int A[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", A[i]);
    printf("\n");
}


int main() {
	int arr[]= {7,5,11,13,6,12};
	printArray(arr,6);
	mergeSort(arr,0,5);
	printArray(arr, 6);
	
	return 0;	
}

File: code/merge_sort.py
def merge(a, l, m, r):
    """Merges 2 subarrays of a[]:
    First subarray is a[l..m]
    Second subarray is a[m+1..r]."""
    n1 = int(m - l + 1) # number of elements in left subarray
    n2 = int(r - m)     # number of elements in right subarray

    L, R = [], []   # create left and right subarrays (temp arrays)

    # copy data to temp arrays
    for i in range(n1):
        L.append(a[l+i])

    for j in range(n2):
        R.append(a[m+1+j])

    i, j = 0, 0 # initial index of left and right subarrays
    k = l       # initial index of merged subarray

    # merge the temp arrays back into a[]
    while i < n1 and j < n2:
        if L[i] <= R[j]:
            a[k] = L[i]
            i += 1
        else:
            a[k] = R[j]
            j += 1
        k += 1

    # copy the remaining elements of L[], if there are any
    while i < n1:
        a[k] = L[i]
        i += 1
        k += 1

    # copy the remaining elements of R[], if there are any
    while j < n2:
        a[k] = R[j]
        j += 1
        k += 1


def merge_sort(a, l, r):
    if l < r:
        m = (l+(r-1))//2      # get middle element
        merge_sort(a, l, m)   # sort left half
        merge_sort(a, m+1, r) # sort right half
        merge(a, l, m, r)     # merge the 2 halves


if __name__ == '__main__':
    a = [5, 7, -29, 0, 178, 2, 4, 11, -2, 6, -9]
    print("Original array: ", a)
    merge_sort(a, 0, len(a)-1)
    print("Sorted array: ", a)
